src/main/java/org/spongepowered/common/world/SpongeTeleportHelper.java

```diff
@@ -47,205 +50,131 @@
 
-        return !(!isBlockSafe(world, blockPos, false) || !isBlockSafe(world, up, false) || !isBlockSafe(world, down, true));
+        // Get the World and get the maximum Y value.
+        int worldMaxY = worldLocation.getExtent().getBlockMax().getY();
 
-    }
+        Vector3i vectorLocation = worldLocation.getBlockPosition();
 
-    private boolean isBlockSafe(World world, Vector3i blockPos, boolean floorBlock) {
-        if (blockPos.getY() <= 0) {
-            return false;
-        }
+        // We use clamp to remain within the world confines, so we don't waste time checking blocks outside of the
+        // world border and the world height.
+        int minY = GenericMath.clamp(vectorLocation.getY() - height, 0, worldMaxY);
+        int maxY = GenericMath.clamp(vectorLocation.getY() + height, 0, worldMaxY);
 
-        if (blockPos.getY() > world.getDimension().getHeight()) {
-            return false;
-        }
-
-        final IBlockState blockState = BlockUtil.getBlockState(world, blockPos);
+        int minX = GenericMath.clamp(vectorLocation.getX() - width, worldBorderMinX, worldBorderMaxX);
+        int maxX = GenericMath.clamp(vectorLocation.getX() + width, worldBorderMinX, worldBorderMaxX);
 
-        if (floorBlock) {
-            // Floor is air so we'll fall, need to make sure we fall safely.
-            if (blockState.getBlock() == BlockTypes.AIR) {
-                final BlockState typeBelowPos = world.getBlock(blockPos.sub(0, 1, 0));
-                final BlockState typeBelowPos2 = world.getBlock(blockPos.sub(0, 2, 0));
+        int minZ = GenericMath.clamp(vectorLocation.getZ() - width, worldBorderMinZ, worldBorderMaxZ);
+        int maxZ = GenericMath.clamp(vectorLocation.getZ() + width, worldBorderMinZ, worldBorderMaxZ);
 
-                // We'll fall too far, not safe
-                if (typeBelowPos.getType() == BlockTypes.AIR && typeBelowPos2.getType() == BlockTypes.AIR) {
-                    return false;
+        // We now iterate over all possible x, y and z positions to get all possible vectors.
+        List<Vector3i> vectors = new ArrayList<>();
+        for (int x = minX; x <= maxX; x++) {
+            for (int y = minY; y <= maxY; y++) {
+                for (int z = minZ; z <= maxZ; z++) {
+                    vectors.add(new Vector3i(x, y, z));
                 }
+            }
+        }
 
-                // We'll fall onto a block, need to make sure its safe
-                if (typeBelowPos.getType() != BlockTypes.AIR && !isSafeFloorMaterial(BlockUtil.toNative(typeBelowPos).getMaterial())) {
-                    return false;
-                }
+        // Sort them according to the distance to the provided worldLocation.
+        return vectors.stream()
+                .sorted((first, second) -> Integer.compare(vectorLocation.distanceSquared(first), vectorLocation.distanceSquared(second)))
+                .collect(Collectors.toList());
```

Comment:
> Returning the stream itself, rather then the collection would be a rather simple fix.

Yep, I'm still thinking in terms of lists, I'll do that.

> In addition, the old method of iterating over the locations was able to generate them roughly in distance order to begin with.

Going back to a modifed original method is something I could do, I just pretty much wrote my own routine and disregarded everything else because I was honestly not sure what was working and what wasn't.

> I'm also slightly worried about a spherical iteration order making it easier to clip into protected areas. Although I think it's impossible to have safe teleport, and not have the user be teleported into areas they arn't supposed to get to, unless the plugin provides some sort of predicate for testing locations.

I suppose it can't harm to add another API method to do that.